internal/ssh/ssh.go

```diff
@@ -0,0 +1,195 @@
+	if host == "" {
+		host = "127.0.0.1"
+	}
+	if port == 0 {
+		port = 22
+	}
+	// Init the SSH config.
+	config := &ssh.ClientConfig{
+		User: user,
+		Auth: []ssh.AuthMethod{
+			ssh.PublicKeys(keypair),
+		},
+		HostKeyCallback: func(hostname string, remote net.Addr, key ssh.PublicKey) error {
+			// If 'hostKeys' was not provided to 'Connect', simply return nil.
+			//
+			// This behavior is the same as 'ssh.InsecureIgnoreHostKey'.
+			if len(hostKeys) == 0 {
+				return nil
+			}
+			// If 'hostKeys' was provided to 'Connect', validate the SSH connection's
+			// host key matches one of 'hostKeys'.
+			for _, hostKey := range hostKeys {
+				if bytes.Equal(hostKey.Marshal(), key.Marshal()) {
+					return nil
+				}
+			}
+			return ErrHostKeyInvalid
+		},
+		Timeout: sshDefaultTimeout,
+	}
+	// Parse the host + port combination to a ssh.Dial-compatible 'addr' (host+
+	// port string).
+	target, err := joinHostPort(host, port)
+	if err != nil {
+		return nil, err
+	}
+	// Dial the SSH connection.
+	client, err := ssh.Dial("tcp", target, config)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %w", ErrSSHFailedDial, err)
+	}
+	return client, nil
+}
+
+// joinHostPort parses and validates 'host' is a valid IPv4 or IPv6 address,
+// then joins it with the port in the address-family-specific format.
+//
+// If 'host' is a hostname, the hostname will be resolved, then hostToPort will
+// recurse using the first of the resolved addresses.
+func joinHostPort(host string, port uint16) (string, error) {
+	// Set up a context for deadline.
+	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+	defer cancel()
+	// Parse and resolve the provided host, join the result with the port as
+	// appropriate.
+	if addr := net.ParseIP(host); addr == nil {
+		// Is it a hostname?
+		addrs, err := net.DefaultResolver.LookupHost(ctx, host)
+		if err != nil {
+			return "", fmt.Errorf("%w: %s", ErrFailedHostParse, host)
+		}
+		// Select the first address we resolved and parse that.
+		return joinHostPort(addrs[0], port)
+	} else if ipv4 := addr.To4(); ipv4 != nil {
+		// 'host' is ipv4
+		return fmt.Sprintf("%s:%d", ipv4.String(), port), nil
+	} else if ipv6 := addr.To16(); ipv6 != nil {
+		// 'host' is ipv6
+		return fmt.Sprintf("[%s]:%d", ipv6.String(), port), nil
+	} else {
+		panic("impossible")
+	}
+}
+
+var (
+	ErrSessionInit     = fmt.Errorf("failed to begin SSH session")
+	ErrCMDExec         = fmt.Errorf("failed to execute SSH command")
+	ErrInWait          = fmt.Errorf("SSH command did not exit cleanly")
+	ErrStdinWrite      = fmt.Errorf("failed to write command to stdin")
+	ErrStdinShortWrite = fmt.Errorf("short write to stdin")
+	ErrStdStreamClose  = fmt.Errorf("encountered error closing standard stream")
+)
+
+// Exec executes a single command, returning any standard out/err received.
+func Exec(client *ssh.Client, cmd string) (string, string, error) {
```

Comment:
stdout/stderr can sometimes be huge things we don't want to buffer, or things that callers want to control. would it make more sense for `Exec` to accept a pair of `io.Writer`s